docs/proposals/proxy-extensions.md

```diff
@@ -0,0 +1,363 @@
+needs to be deployed in each of the target clusters. To address this
+use-case there is a need to configure multiple backend URLs for the
+same extension (one for each cluster). For doing so, the following
+configuration should be possible:
+
+```yaml
+extension.config: |
+  extensions:
+    - name: some-extension
+      enabled: true
+      backend:
+        idleConnTimeout: 10s
+        services:
+          - url: http://extension-name.com:8080
+            clusterName: kubernetes.local
+          - url: https://extension-name.ppd.cluster.k8s.local:8080
+            clusterName: admins@ppd.cluster.k8s.local
+```
+
+Note that there is an URL configuration per cluster name. The cluster
+name is extracted from the Argo CD cluster secret and must match the
+field `data.name`. In this case the UI must send a header with the
+application name so API Server can check in which cluster it should
+get the backend URL from. This will be done by inspecting the
+Application destination configuration to find the proper cluster
+name.
+
+The diagram below shows how Argo CD UI could send the request with
+the additional header to get the proxy forwarding it to the proper
+cluster:
+
+```
+   ┌────────────┐
+   │ Argo CD UI │
+   └──────┬─────┘
+          │
+          │ GET http://argo.com/api/v1/extensions/some-extension
+          │ HEADER: "ARGOCD-APPLICATION-NAME: ppd-application
+          │
+          ▼
+ ┌──────────────────┐
+ │Argo CD API Server│
+ └────────┬─────────┘
+          │
+          │ GET https://extension-name.ppd.cluster.k8s.local:8080
+          │
+          ▼
+  ┌───────────────┐
+  │Backend Service│
+  └───────────────┘
+```
+
+##### Considerations
+
+- The `idleConnTimeout` can be used to avoid accumulating too many
+  goroutines waiting slow for extensions. In this case a proper timeout
+  error (408) should be returned to the browser.
+- Headers, scheme, http verb and request body are forwarded as it is
+  received by the API server to the backend service.
+- A new header is added in the forwared request (`X-Forwarded-Host`) to
+  allow ssl redirection.
+- This proposal doesn't specify how backends should implement authz or
+  authn. This topic could be discussed as a future enhancement to the
+  proxy extension feature in Argo CD.
+
+----
+
+#### [UC-2]: As an Argo CD admin, I want to define extensions rbacs so access permissions can be enforced
+
+Extend Argo CD rbac registering a new `ResourceType` for extensions in the
+[policy configuration][3]. The current policy permission configuration is
+defined as:
+
+```
+p, <subject>, <resource>, <action>, <object>, <access>
+```
+
+With a new resource type for extensions, admins will be able to configure
+access rights per extension per project. The `object` field must contain
+the project name and the extension name in the format
+`<project>/<extension>`
+
+- **Example 1**:
+
+```
+p, role:allow-extensions, extensions, *, some-project/some-extension, allow
```

Comment:
What are the `actions` for this resource?

I'd suggest that we provide the HTTP verbs as actions, i.e. `POST, GET, PUT, PATCH, and DELETE`.

Reply:
I thought about that and we discussed a bit in the contrib's meeting when I presented. While the verbs makes sense I personally think that we should provide a simpler first version and improve if community see the necessity. With the current RBAC config admins would be able to enable/disable extensions in a given project and maybe this is good enough?

Reply:
Makes sense! To rephrase, in v1 the _only_ acceptable value in the CSV's `action` field will be an asterisk. Later, when we define what `action` means for an `extensions` resource, everyone's existing RBAC config will behave the same. They'll have the option to further-restrict their RBAC if they want.

Reply:
Am I correct in saying that there is no way to configure an extension per application rbac?

Reply:
Had the same [thought](https://github.com/argoproj/argo-cd/pull/10435#discussion_r986941880). :-)

Reply:
> Am I correct in saying that there is no way to configure an extension per application rbac?

Not per application but per project.

Reply:
@crenshaw-dev added in the Q&A regarding our discussion. LGTY?